org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java

```diff
@@ -346,25 +362,29 @@ public TypeBinding resolveType(BlockScope scope) {
 		if ((expressionType != TypeBinding.NULL && expressionType.isBaseType()) // disallow autoboxing
 				|| checkedType.isBaseType()
 				|| !checkCastTypesCompatibility(scope, checkedType, expressionType, null, true)) {
-			checkForPrimitives(scope, checkedType, expressionType);
+			checkForPrimitives(scope, checkedType, expressionType, true);
 		}
 	}
 
 	return this.resolvedType = TypeBinding.BOOLEAN;
 }
 
-private void checkForPrimitives(BlockScope scope, TypeBinding checkedType, TypeBinding expressionType) {
+private void checkForPrimitives(BlockScope scope, TypeBinding checkedType, TypeBinding expressionType, boolean flagError) {
 	PrimitiveConversionRoute route = Pattern.findPrimitiveConversionRoute(checkedType, expressionType, scope);
 	this.testContextRecord = new TestContextRecord(checkedType, expressionType, route);
 
 	if (route == PrimitiveConversionRoute.WIDENING_PRIMITIVE_CONVERSION
 			|| route == PrimitiveConversionRoute.NARROWING_PRIMITVE_CONVERSION
 			|| route == PrimitiveConversionRoute.WIDENING_AND_NARROWING_PRIMITIVE_CONVERSION) {
-//				this.expression.computeConversion(scope, expressionType, checkedType);
-	} else if (route == PrimitiveConversionRoute.NO_CONVERSION_ROUTE) {
-		scope.problemReporter().notCompatibleTypesError(this, expressionType, checkedType);
-	} else {
+
+		// Do Nothing - no additional steps required for conversion later.
+
+	} else if (route == PrimitiveConversionRoute.BOXING_CONVERSION
+			|| route == PrimitiveConversionRoute.BOXING_CONVERSION_AND_WIDENING_REFERENCE_CONVERSION) {
 		addSecretExpressionValue(scope, expressionType);
+	} else if (route == PrimitiveConversionRoute.NO_CONVERSION_ROUTE) {
+		if (flagError)
+			scope.problemReporter().notCompatibleTypesError(this, expressionType, checkedType);
```

Comment:
I find it difficult to see, why some invocations should check these conversions and not report an error. I'll propose a small change that avoids calling checkForPrimitives() if no primitives are in sight, which then allows removing `flagError`.